diff --git a/packager/react-native-xcode.sh b/packager/react-native-xcode.sh
index d136282fd..d74024637 100755
--- a/packager/react-native-xcode.sh
+++ b/packager/react-native-xcode.sh
@@ -91,7 +91,7 @@ $NODE_BINARY "$REACT_NATIVE_DIR/local-cli/cli.js" bundle \
   --dev $DEV \
   --reset-cache \
   --bundle-output "$BUNDLE_FILE" \
-  --assets-dest "$DEST"
+  --assets-dest "$DEST" --customExtensions=e2e

 if [[ ! $DEV && ! -f "$BUNDLE_FILE" ]]; then
   echo "error: File $BUNDLE_FILE does not exist. This must be a bug with" >&2
diff --git a/react.gradle b/react.gradle
index 09c06ca7e..f7a73dadd 100644
--- a/react.gradle
+++ b/react.gradle
@@ -79,10 +79,10 @@ gradle.projectsEvaluated {
                 def devEnabled = !targetName.toLowerCase().contains("release")
                 if (Os.isFamily(Os.FAMILY_WINDOWS)) {
                     commandLine("cmd", "/c", *nodeExecutableAndArgs, "node_modules/react-native/local-cli/cli.js", "bundle", "--platform", "android", "--dev", "${devEnabled}",
-                            "--reset-cache", "--entry-file", entryFile, "--bundle-output", jsBundleFile, "--assets-dest", resourcesDir, *extraPackagerArgs)
+                            "--reset-cache", "--entry-file", entryFile, "--bundle-output", jsBundleFile, "--assets-dest", resourcesDir, "--customExtensions", "e2e", *extraPackagerArgs)
                 } else {
                     commandLine(*nodeExecutableAndArgs, "node_modules/react-native/local-cli/cli.js", "bundle", "--platform", "android", "--dev", "${devEnabled}",
-                            "--reset-cache", "--entry-file", entryFile, "--bundle-output", jsBundleFile, "--assets-dest", resourcesDir, *extraPackagerArgs)
+                            "--reset-cache", "--entry-file", entryFile, "--bundle-output", jsBundleFile, "--assets-dest", resourcesDir, "--customExtensions", "e2e", *extraPackagerArgs)
                 }

                 enabled config."bundleIn${targetName}" ||
